sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala

```diff
@@ -99,6 +99,16 @@ abstract class JdbcDialect extends Serializable with Logging{
    */
   def getJDBCType(dt: DataType): Option[JdbcType] = None
 
+  /**
+   * Get the factory method for create JDBC connection.
+   * In general, creating a connection has nothing to do with JDBC partition.
+   * But sometimes it is needed, such as a database with multiple shard nodes.
+   * @param options JDBC options.
+   * @return The factory method for create JDBC connection.
+   */
+  def getConnection(options: JDBCOptions): JDBCPartition => Connection =
```

Comment:
This is a developer-facing API, do we really want to expose `JDBCPartition` to users?

Reply:
Yes. Users could use the JDBCPartition to get connection with different partition.

Reply:
what information do we need exactly? If you need partition id, isn't `Int` good enough here?

Reply:
Yes. we need partition id.